Split evolution initialization

nilsdeppe opened the pull request:
## Proposed changes

- Fixes a bug where python bindings code was built when it wasn't needed to be.
- Adds a function to insert things that aren't present in the DataBox that is used in initialization.
- Fixes a bug in `ActionTesting::tag_is_retrievable` (and adds the test I should've written)
- Adds metafunctions that compute what needs to be added for the initialization phase
- Splits the `InitializeElement` simple action into several iterable actions
- Changes Burgers and ScalarWave to use the new initialization actions
- Add ability to disable flux boundary conditions
- Add ability to specify which quantities to slice to interfaces and what additional compute items should be on the interface.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add function to optionally add something to DataBox

This will be used in initialization code.

nilsdeppe committed:
Add has_analytic_solution_alias type trait

nilsdeppe committed:
Factor Interface initialization

The interface initialization now takes as template parameters additional tags
that are to be sliced to the face and exteriors, as well as compute tags for the
face and exteriors. This should allow all systems to share the Interface
initialization action.

nilsdeppe committed:
Add char speeds for Burgers

nilsdeppe committed:
Use new initialization in EvolveBurgers

nilsdeppe committed:
Disable another python bindings lib when bindings are off

nilsdeppe committed:
Add nonconservative system initialization action

nilsdeppe committed:
Use new initialization actions in EvolveScalarWave

nilsdeppe committed:
Remove InitializeElement action

nilsdeppe committed:
Provide option to disable flux boundary conditions

nilsdeppe committed:
Add Initialization doxygen group

nilsdeppe committed:
Fix bug in ActionTesting::tag_is_retrievable

Previously this would incorrectly return true.

nilsdeppe committed:
Enable initialization of conservative systems w/o prims

nilsdeppe committed:
Add metafunctions and actions for handling options in init

nilsdeppe committed:
Split evolution initialization into separate actions

nilsvu commented on src/Evolution/TypeTraits.hpp:
```diff
@@ -0,0 +1,26 @@
+
+#include "Utilities/TypeTraits.hpp"
+
+namespace evolution {
```
I don't think this namespace exists yet, so please document it somewhere so that it shows up in Doxygen. I expect it will see a lot more use :)

Also, I'd prefer CamelCase for it (`Evolution`) since it's not an abbreviation. Then it also matches the existing CamelCase `Elliptic` namespace. If you strongly prefer lower-case for some reason, then please also rename the `Elliptic` namespace to lower-case. I know we haven't established strict rules for namespace names, but I think at least these similar namespaces should match somehow.

nilsdeppe replied:
I'll change the elliptic namespace to snake case.

nilsvu replied:
If you want to make that change, please give some reasoning

nilsdeppe replied:
lower case letters are easier to type

nilsvu replied:
that‘s fair :) looks like we are ending up with lower-case, mostly abbreviated namespaces at the top level, and slightly more verbose CamelCase namespaces at lower levels. We could make that a guideline I guess.

nilsdeppe replied:
I don't think the elliptic namespace is documented either. Or at least it's not showing up.

nilsvu commented on src/Evolution/TypeTraits.hpp:
```diff
@@ -0,0 +1,26 @@
+/// Inherits from `std::true_type` if `typename T::analytic_solution` is
+/// well-formed.
+template <class T, class = cpp17::void_t<>>
+struct has_analytic_solution_alias : std::false_type {};
```
Having an analytic solution is not limited to evolution systems, so can you move this to a more general place? E.g. `PointwiseFunctions/AnalyticSolutions/TypeTraits.hpp` or `Utilities/TypeTraits.hpp`

nilsdeppe replied:
Sure, someone else can move it in the future. See my general comment.

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,117 @@
+namespace Initialization {
+namespace detail {
+template <typename AddingAction, typename ComputeTagsList,
+          bool OverwriteSimpleVars, typename... SimpleTags, typename DbTagsList,
```
`OverwriteSimpleTags` here and below

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,117 @@
+                          db::item_type<typename decltype(tag_v)::type>>> {
+        using tag = typename decltype(tag_v)::type;
+        if (db::get<tag>(box) != tuples::get<tag>(simple_tags)) {
+          ERROR("While not add the tag "
```
something about this sentence is incorrect

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,117 @@
+        Parallel::printf(
+            "Note: Unable to check that the initial value of tag '%s' being "
+            "re-added is the same as that in the DataBox. The action re-adding "
+            "the tag is '%s'\n.",
```
I think this should `ERROR` out, since that would be the safe thing to do.

However, if you don't want this case to be an error, then we have to assume the user does it intentionally (has read the documentation) and doesn't want to be alerted of a "potential issue" every time the code is run, so this should be silent at runtime.

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,117 @@
+}  // namespace detail
+
+/*!
+ * Adds tags that are not yet in the DataBox.
```
is `\brief` not necessary?

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,117 @@
+template <typename AddingAction, typename ComputeTagsList,
+          bool OverwriteSimpleVars = false, typename... SimpleTags,
+          typename DbTagsList>
+auto add_nonexistent_to_databox(
```
This is fairly general functionality, so can you move this to `DataBox`? At least, it's not evolution- or (strictly) initialization-specific. How about merging this into `db::create_from`, by adding a template parameter `IgnoreExistingTags` that defaults to `false`?

nilsdeppe replied:
I don't think it should be used outside of initialization. I'm worried it'll get people into the bad habit of rather than understanding and fixing a compiler error, just saying "Meh, ignore duplicates". I'm really not open to moving this for that reason. I thought about it a lot while writing it and I think it's a bad decision.

nilsvu replied:
Agreed. How about naming this `merge_into_databox`?

nilsdeppe replied:
I'm not sure, need to think about it more :)

nilsdeppe replied:
After more thought and discussion I agree with this.

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -48,39 +72,54 @@ namespace Actions {
+template <typename System, typename SliceTagsToFace,
+          typename SliceTagsToExterior,
+          typename FaceComputeTags = face_compute_tags<>,
+          typename ExteriorComputeTags = exterior_compute_tags<>>
```
I agree that we should allow more flexibility in what system-specific things are sliced to the interface. However it's probably not the `Metavariables` that has to decide that for each executable individually, but it just depends on the kind of system or the numerical scheme which tags are needed. So I think we don't really need this functionality where a `Metavariables` "inserts" tags into the interface initializer, but rather we have more fine-grained initializer actions do that.

More specifically, I suggest:
- Have an initializer that just adds directions, and simple interface things like face-meshes
- Then, have a system initializer that adds everything it needs, including slicing whatever it wants to the faces and adding what is needed to compute magnitudes
- Then, have an initializer that adds face normals
- Each system or whatever component of our code is free to provide more init actions, they can just be added to the action list. We don't have to add all that functionality into a few big init actions anymore.

In particular, most of the executables slice the same things to the interface because they use the same numerical scheme. So just have a dedicated initializer that adds the primitive vars and so forth.

nilsdeppe replied:
I don't agree with any of what you said. I think the issue is more subtle because like you said, it depends on the scheme, etc. and **also** on the order that things need to be added. Again, I'm making the minimal change to allow Burgers, scalar wave, GRMHD, and (hopefully) Newtonian Euler to use the same code.

nilsvu replied:
I just fear this will be abused and users add long lists of sliced variables to the Metavariables instead of writing another initializer..

nilsdeppe replied:
In my opinion that isn't anymore abusive than having a `AddComputeItems` action. What you're calling abuse is what I sort of view as the intended behavior.

nilsvu replied:
we decided to keep the current design for now and later split off initializers for numerical fluxes

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,71 @@
+/// Available actions to be used in the initialization.
+namespace Actions {
+template <typename OptionsTagsToRemove>
+struct RemoveOptionsFromDataBox {
```
This action really just removes tags from the DataBox, so it's not evolution- or even initialization-specific. I suggest to name it `RemoveTags` and move it to a more general place, basically where you would like DataBox-actions like `Actions::MutateApply` to end up. If I understood correctly, you don't like to have a `DataStructures/DataBox/Actions` directory. Then perhaps `Utilities/Actions`?

nilsdeppe replied:
There should be exactly zero actions in `Utilities` and `DataStructures`. I'm not sure where they should go, but definitely not in those directories. `NumericalAlgoritms/Actions` is an option, but not super great. Another place in the same situation is `Parallel/Actions` since these actions aren't really related to the parallel code. Again, minimal changes so I'm not really open to moving it.

nilsvu replied:
we (tentatively) decided to move these things to `/ParallelAlgorithms/Initialization/` (possibly in another PR)

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,71 @@
+/// terminates the phase. None of the initialization actions should terminate
+/// the algorithm.
+template <typename ActionsList>
+using generate_action_list =
```
I suggest that instead of adding the wrappers `generate_add_options_to_databox` and `generate_action_list` you just expose your `detail::options_databox_tags` metafunction, naming it `collect_option_tags` for instance. Then we can just use `Parallel::ForwardAllOptionsToDataBox` and `Actions::RemoveTags` (if you do the renaming I suggested above) directly in the Metavariables, instead of having these wrappers.

Reasoning: I'm generally in favour of not adding wrappers if the thing they are wrapping is easier to do directly than learning how to use the wrapper. In this particular case, most value is gained from wrapping the call to `detail::options_databox_tags<ActionsList>`, and if we just expose that I think we don't need the wrappers anymore.

nilsdeppe replied:
This might be okay for `generate_add_options_to_databox`, but not for `generate_action_list` since there specific actions are added. Please keep in mind that telling someone to use `generate_action_list<Actions...>` is easier for them than looking at the `tmpl::append` code. This is **user interface** code, and so the goal should be to make it not scary to use and easy to read.

nilsvu replied:
I agree we should make the interface nice for the user! Indeed I think it's actually more transparent and understandable for the user to add `Parallel::ForwardAllOptionsToDataBox` somewhere, and then in the initialization action list explicitly add a `RemoveOptionsFromDatabox` and then a `TerminatePhase`. It's quite nice that the PDALs quite explicitly tell the story what the program will do, and I'd like to avoid hiding that in wrappers. Thoughts?

nilsdeppe replied:
I'm worried people will forget about the `RemoveOptionsFromDatabox` and `TerminatePhase`, which will lead to extra things in the DataBox for no reason, and possibly inexplicable infinite loops at runtime. I'd rather prevent those two bugs from appearing at all.

nilsvu replied:
I understand your worry. I‘d nevertheless prefer the transparency, since users must know how action lists work for writing an executable.. I‘ll think about it a bit more, we can talk tomorrow

nilsvu replied:
We decided to have a `RemoveOptionsAndTerminatePhase` action that takes no template parameters, so removes all options. We can also expose `RemoveOptions<OptionTags...>` on its own. Also expose `collect_option_tags` so that `Forward...<>` can be used explicitly in the metavariables.

nilsvu commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -0,0 +1,48 @@
+namespace Tags {
+
+template <size_t Dim>
+struct InitialExtents : db::SimpleTag {
```
Can the two domain-related tags here go some that's not evolution-specific?

nilsdeppe replied:
In theory yes, in practice, this PR is just for evolution code.

nilsvu commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -20,8 +32,9 @@ struct Inertial;
 namespace Initialization {
-
-/// \brief Allocate variables needed for evolution of conservative systems
+namespace Actions {
```
I suggest to remove the outer `Initialization` namespace and instead place the evolution-specific initializers in `Evolution::Actions`, e.g. `Evolution::Actions::InitializeConservativeSystem`. Reason is that initializers are regular actions now, so we should treat them as such. Also, actions should have verbs for names.

nilsdeppe replied:
Meh, I'm not sure about the evolution renaming. I'm not for or against, I'd rather leave it as is.

As I've said before, I'm not "namespacing" all the initialization actions with `Initialize`. They should just go in an `Initialization` namespace.

nilsvu replied:
Then, how about we swap the order of nested namespaces: `Actions::Initialize::ConservativeSystem` or `Actions::Initialization::ConservativeSystem`
Alternatively, we could skip the `Actions` namespace for initializers and do `Initializers::ConservativeSystem` or `InitializeActions::ConservativeSystem`

nilsvu replied:
we decided to do names like `Actions::InitializeDomain`

nilsdeppe replied:
In a future PR :)

nilsvu commented on src/Evolution/Initialization/DiscontinuousGalerkin.hpp:
```diff
@@ -30,10 +33,14 @@
 
 namespace Initialization {
-
+namespace Actions {
```
I suggest to name this `dg::Actions::InitializeMortars` and place it in `NumericalAlgorithms/DG/Actions`

nilsdeppe replied:
You are welcome to do this in a PR later.

nilsvu replied:
sure 👍

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -19,18 +20,27 @@
 
 namespace Initialization {
-
+namespace Actions {
```
I suggest to name this `(domain::)Actions::InitializeDomain` and place it in `Domain/Actions/`

nilsdeppe replied:
I'm not sure how I feel about `Domain/Actions`, but it shouldn't be part of this PR.

nilsvu replied:
sure 👍

nilsvu commented on src/Evolution/Initialization/Helpers.hpp:
```diff
@@ -16,7 +16,6 @@
 
 /// Items for initializing the DataBox%es of parallel components
 namespace Initialization {
-
```
This function can probably be merged into the domain initializer

nilsdeppe replied:
That's no longer `minimal`. I think it's clear that after this PR the initialization will still need some work to get it in better shape. This sort of change should be part of a **series** of PRs that do that.

nilsvu replied:
sure 👍

nilsvu reviewed (changes requested):
Not a complete review yet, but some suggestions on how we could structure initializers

nilsdeppe commented:
@nilsleiffischer while I agree with many of your suggestions in terms of long-term unifying as much of the elliptic and evolution initialization, that is not what this PR is doing. This PR is making the minimal amount of changes to the evolution initialization necessary to have all the evolution executable use the same initialization code. Anything more should be done in a separate PR. I'll respond to individual comments with my thoughts as well :)

nilsvu reviewed (commented):
Sure, we (i.e., I) can move things out of Evolution to more general places later. Minimal changes sounds good :) Focusing on the init code you are adding in this PR then

nilsvu reviewed (commented):
notes from a call with @nilsdeppe and myself

nilsdeppe committed:
Rename Elliptic namespace to elliptic

nilsdeppe commented:
@nilsleiffischer I've pushed fixups and updates to this. Could you double check to make sure I didn't miss anything we discussed and that there wasn't any disagreement during our discussion? Thanks :)

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+          typename DbTagsList,
+          Requires<Policy == MergePolicy::Overwrite> = nullptr>
+void check_simple_tag_value(const gsl::not_null<db::DataBox<DbTagsList>*> box,
+                            db::item_type<SimpleTag>&& simple_tag) noexcept {
```
`simple_tag_value`

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+template <typename AddingAction, typename SimpleTag, MergePolicy Policy,
+          typename DbTagsList,
+          Requires<Policy == MergePolicy::Overwrite> = nullptr>
+void check_simple_tag_value(const gsl::not_null<db::DataBox<DbTagsList>*> box,
```
perhaps `merge_simple_tag_value` or `resolve_simple_tag_value`

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+                            db::item_type<SimpleTag>&& simple_tag) noexcept {
+  db::mutate<SimpleTag>(
+      box,
+      [](const auto item_ptr,
```
`auto` -> `gsl::not_null<db::item_type<SimpleTag>*>` ?
`item_ptr` -> `stored_simple_tag_value`

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+  db::mutate<SimpleTag>(
+      box,
+      [](const auto item_ptr,
+         db::item_type<SimpleTag>&& local_simple_tag) noexcept {
```
`new_simple_tag_value`?

nilsdeppe replied:
We've been using the `local_` prefix to avoid GCC's false positive warning in other places so this seems consistent to me.

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+void check_simple_tag_value(const gsl::not_null<db::DataBox<DbTagsList>*> box,
+                            db::item_type<SimpleTag>&& simple_tag) noexcept {
+  (void)box;
+  (void)simple_tag;
```
can you not just comment out the function arguments?

nilsdeppe replied:
Yes, I previously had a runtime error so these were used...

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -15,59 +15,82 @@
+template <typename AddingAction, typename ComputeTagsList, MergePolicy Policy,
+          typename... SimpleTags, typename DbTagsList,
           typename... SimpleTagsToAdd, typename... SimpleTagsToCheck>
 auto add_nonexistent_to_databox_impl(
```
`merge_into_databox_impl`

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -84,7 +107,7 @@ auto add_nonexistent_to_databox_impl(
 
 /*!
- * Adds tags that are not yet in the DataBox.
+ * \brief Adds tags that are not yet in the DataBox.
```
`Adds` -> `Add`

Please update this documentation with the changes you made in this fixup

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -95,9 +118,9 @@ auto add_nonexistent_to_databox_impl(
+          MergePolicy Policy = MergePolicy::Error, typename... SimpleTags,
           typename DbTagsList>
-auto add_nonexistent_to_databox(
+auto merge_into_databox(
```
What will happen if a tag is retrievable from the databox, but it is not stored as one of the tags that being added here? E.g. the tag is already computed by a compute item or stored in a derived tag. Can we print a useful error message and add a test?

nilsdeppe replied:
- You will get an error when you try to retrieve it.
- We already do.
- No, we can't test static asserts.

nilsvu commented on tests/Unit/Evolution/Initialization/Test_AddToDataBox.cpp:
```diff
@@ -38,72 +38,66 @@ struct TagMultiplyByTwo : db::ComputeTag {
-      Initialization::add_nonexistent_to_databox<FakeAction, tmpl::list<>>(
-          std::move(box0), tuples::TaggedTuple<TagInt>{2});
+  auto box1 = Initialization::merge_into_databox<FakeAction, tmpl::list<>>(
+      std::move(box0), tuples::TaggedTuple<TagInt>{2});
```
You don't _have to_ chain _all_ of these calls, you could also set up a box and then `SECTION` or scope the following calls so that they are more self-contained tests. Up to you tho

nilsdeppe replied:
the code will likely double in size, if not more, with `SECTION`s, so I think this is actually cleaner.

nilsvu commented on src/Evolution/Evolution.hpp:
```diff
@@ -0,0 +1,10 @@
+
+#pragma once
+
+/// Classes, functions, etc. used by the evolution part of SpECTRE.
```
`Functionality for evolving hyperbolic PDEs`?

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -91,7 +91,7 @@ struct Domain {
                                          my_block.coordinate_map().get_clone()};
 
-    return std::make_tuple(add_nonexistent_to_databox<Domain, compute_tags>(
+    return std::make_tuple(merge_into_databox<Domain, compute_tags>(
```
Can you make the `Action` template parameter a meta argument instead or move it after the simple and compute tags list template parameters, so that it doesn't break the nice `merge_into_databox<the_tags_to_merge>` semantics of reading this?

nilsdeppe replied:
In theory yes, it makes the code harder to use in my opinion. I'm very strongly against this suggestion.

Edit: I'm strongly against making it a `meta` argument. Swapping the order will be a huge pain in the butt in terms of rebasing...

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -91,7 +91,7 @@ struct Domain {
+    return std::make_tuple(merge_into_databox<Domain, compute_tags>(
         std::move(box),
         tuples::tagged_tuple_from_typelist<simple_tags>{
             std::move(mesh), std::move(element), std::move(map)}));
```
I think it would be cleaner to skip the conversion to the tagged tuple and just pass the simple tags values as arguments, as is done in `db::create`

nilsdeppe replied:
Can't be done in a safe manner. The only other option might be `merge_into_databox<Domain, simple_tags, compute_tags>(std::move(box), std::move(BLAH)...)`.

Edit: I have some concern with that about people passing in arguments in the correct order. It should be fine but might not.

nilsleiffischer replied:
I like it because it's what `db::create` does

nilsvu replied:
I like it because it's what `db::create` does

nilsvu reviewed (changes requested):
Please squash the fixups

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -25,10 +25,35 @@ class ConstGlobalCache;
 namespace Initialization {
+/// %Tags that are to be sliced to the faces of the element
+template <typename... Tags>
+using slice_tags_to_face = tmpl::list<Tags...>;
```
`slice_to_faces` and `slice_to_exteriors`?

nilsdeppe replied:
I prefer the current

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -25,10 +25,35 @@ class ConstGlobalCache;
+
+/// Compute tags on the faces of the element
+template <typename... Tags>
+using face_compute_tags = tmpl::list<Tags...>;
```
`compute_on_faces` and `compute_on_exteriors`?

nilsdeppe replied:
I prefer the current

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -48,39 +72,54 @@ namespace Actions {
+                          typename System::variables_tag>>;
+
+  template <typename TagToSlice, typename Directions>
+  struct make_slice_tag {
```
These should probably be private (everything except `apply`)

nilsdeppe replied:
Leaving as is since we don't care in other parts of t he code.

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,66 @@
+template <typename Action>
+struct get_options_tags<
+    Action, cpp17::void_t<typename Action::initialization_option_tags>> {
+  using type = typename Action::initialization_option_tags;
```
I think the alias can be just `option_tags` (instead of `initialization_option_tags`) since also for parallel components the `option_tags` alias is just for initialization

nilsdeppe replied:
Yes, but those should be kept separate.

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,66 @@
+
+/// Get the list of input file options from the list of initialization actions.
+template <typename InitializationActionList>
+using option_tags = tmpl::remove_duplicates<tmpl::flatten<tmpl::transform<
```
`collect_option_tags` is more in line with `observers::collect_reduction_data_tags`

nilsdeppe replied:
:man_shrugging: I like this.

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,66 @@
+    using options_tags_to_remove = option_tags<ActionList>;
+    using options_tags_to_remove_this_time = tmpl::filter<
+        options_tags_to_remove,
+        tmpl::bind<tmpl::list_contains, tmpl::pin<DbTagsList>, tmpl::_1>>;
```
what do lines 57-59 mean/do? could you make that clearer in the code?

nilsdeppe replied:
I'm not sure what you want. A comment? It filters tags from `option_tags_to_remove` that are in `DbTagsList`. That is about as clear as it can get...

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,66 @@
+///
+/// \snippet Evolution/Initialization/Test_Initialize.cpp actions
+///
+/// The `add_options_to_databox` will typically just be:
```
`The parallel component's add_option_to_databox type alias will typically just be`

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -0,0 +1,66 @@
+/// Available actions to be used in the initialization.
+///
+/// The action list for initialization must end with the
+/// `Initialization::Actions::RemoveOptionsAndTerminatePhase` action. For
```
You don't currently enforce this in the action, add an assert?

nilsvu commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -35,55 +48,122 @@ namespace Initialization {
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent,
+            Requires<tmpl::list_contains_v<
+                typename db::DataBox<DbTagsList>::simple_item_tags,
```
Better use `tag_is_retrievable<InitialTime>` since that's all we care about for dependencies

nilsdeppe replied:
We know `InitialTime` is a simple tag, it has to be. Leaving as is.

nilsvu commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -35,55 +48,122 @@ namespace Initialization {
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    return {std::move(box)};
```
This should error out, complaining that the dependencies are not fulfilled. I suggest:
```cpp
ERROR(
    "Dependencies not fulfilled. Did you call the actions in the right "
    "order?");
```

You can also be more specific and list which dependencies are missing

nilsvu commented on src/Evolution/Initialization/DiscontinuousGalerkin.hpp:
```diff
@@ -119,24 +127,23 @@ struct DiscontinuousGalerkin {
-        std::move(box), std::move(mortar_data),
-        std::move(mortar_next_temporal_ids), std::move(mortar_meshes),
-        std::move(mortar_sizes));
+    return add_nonexistent_to_databox<DiscontinuousGalerkin, tmpl::list<>>(
```
The compute tags list should probably default to empty for the merge function

nilsdeppe replied:
I intentionally didn't do that here.

nilsvu commented on src/Evolution/Initialization/DiscontinuousGalerkin.hpp:
```diff
@@ -181,85 +188,92 @@ struct DiscontinuousGalerkin {
-      const std::vector<std::array<size_t, dim>>& initial_extents) noexcept {
-    return impl::initialize(std::move(box), initial_extents);
+  template <typename DbTagsList, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent>
```
You should probably check for the `InitialExtents` dependency (see conservative system)

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -46,32 +56,61 @@ namespace Initialization {
+  using simple_tags = db::AddSimpleTags<::Tags::Mesh<Dim>, ::Tags::Element<Dim>,
+                                        ::Tags::ElementMap<Dim>>;
+
+  using compute_tags = db::AddComputeTags<
```
these typelists can/should now be private, perhaps moved into `apply`

nilsdeppe replied:
Again, we don't do this super consistently throughout the code

nilsvu commented on src/Evolution/Initialization/DiscontinuousGalerkin.hpp:
```diff
@@ -181,85 +188,92 @@ struct DiscontinuousGalerkin {
 
   using impl = Impl<typename Metavariables::system>;
   using simple_tags = typename impl::simple_tags;
   using compute_tags = typename impl::compute_tags;
```
I don't think you need these tags lists anymore. Instead, request `InitialExtents` for `option_tags`

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -46,32 +56,61 @@ namespace Initialization {
+            typename ParallelComponent,
+            Requires<tmpl::list_contains_v<
+                typename db::DataBox<DbTagsList>::simple_item_tags,
+                Tags::InitialExtents<Dim>>> = nullptr>
```
same comment about `db::tag_is_retrievable`

nilsdeppe replied:
Same reply.

nilsvu commented on src/Evolution/Initialization/Domain.hpp:
```diff
@@ -46,32 +56,61 @@ namespace Initialization {
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    return {std::move(box)};
```
same comment about erroring out if the dependencies are not fulfilled

nilsvu commented on src/Evolution/Initialization/Evolution.hpp:
```diff
@@ -49,33 +51,39 @@ namespace Initialization {
+  using initialization_option_tags =
+      tmpl::list<Tags::InitialTime, Tags::InitialTimeDelta,
+                 Tags::InitialSlabSize>;
+
```
everything but the option tags and `apply` should be private

nilsdeppe replied:
We've never be strict on this, so I'm leaving it.

nilsvu commented on src/Evolution/Initialization/Evolution.hpp:
```diff
@@ -136,10 +154,26 @@ struct Evolution {
-        std::move(box), TimeId{}, time_id, initial_dt, std::move(dt_vars),
-        std::move(history));
+    return std::make_tuple(
+        db::create_from<db::RemoveTags<>, simple_tags, compute_tags>(
```
also merge here, instead of `create_from`?

nilsvu replied:
ping

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -41,67 +51,76 @@ namespace Initialization {
-  using simple_tags =
-      db::AddSimpleTags<Tags::Interface<Tags::BoundaryDirectionsExterior<dim>,
-                                        typename System::variables_tag>>;
+  using simple_tags = db::AddSimpleTags<::Tags::Interface<
```
same comments about moving these tags lists into `apply` or make private

nilsdeppe replied:
Leaving as is.

nilsvu commented on src/Evolution/Initialization/Interface.hpp:
```diff
@@ -41,67 +51,76 @@ namespace Initialization {
-    const auto& mesh = db::get<Tags::Mesh<dim>>(box);
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
```
Check for dependencies

nilsdeppe replied:
There are none...

nilsvu replied:
This needs domain things. Now I'm confused, how does this compile when it can't be sure `Tags::Mesh` is in the DataBox? Why don't you have to add a `Requires` so that only when the incoming DataBox contains the dependencies will this `apply` run? `merge_into_databox` doesn't protect against missing dependencies, or did I miss something?
Unresolving this comment just so that I understand what's going on.

nilsdeppe replied:
This doesn't depend on any of the initialization options. This action behaves the same as pretty much every other action in the code.

nilsvu replied:
It's in the init phase though, so it's DataBox has to work with _all_ DataBoxes that come out of the other actions in this phase. Does this work because the domain initializer is the _first_ action in the list, so the Interface initializer is guaranteed to have the domain stuff in the DataBox?

nilsdeppe replied:
Oh, I see your confusion now. Yes, because you cannot have interfaces without a domain, the domain is initialized first and so this works. Does that make sense?

nilsvu replied:
Yes. If the domain was initialised, say, second in the init action list, then I think we would get compiler errors because it's not guaranteed that the actions are run in order.

nilsvu replied:
So I think the initialisers have to check their dependencies. We can't rely on the domain initializer always being the very first in the list.

nilsvu replied:
Maybe I am still getting something wrong, actually. The interface initializer slices variables to the faces, but how can it be sure the system initializer was run? In principle the algorithm could jump from the domain action directly to the interface action..

nilsdeppe replied:
My personal view here is that at some point we need users to think about what they're doing. Trying to safeguard against everything is impossible, and will drive compile time through the roof. It's these types of situations where I think we should use documentation (which I'm happy to add here) and examples.

nilsvu commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp:
```diff
@@ -15,242 +15,77 @@
-          std::move(box), std::move(gr_vars));
-    }
-  };
+struct GrTags {
```
name `InitializeGrTags`

nilsvu commented on src/Evolution/Initialization/NonconservativeSystem.hpp:
```diff
@@ -0,0 +1,84 @@
+    using solution_tag = ::OptionTags::AnalyticSolutionBase;
+    Vars vars{num_grid_points};
+    vars.assign_subset(Parallel::get<solution_tag>(cache).variables(
+        inertial_coords, initial_time, typename Vars::tags_list{}));
```
use `db::get_variables_tags_list`, which works also for prefixed variables, if only to save someone who copies this code from a potential bug

nilsdeppe replied:
But here having a prefix would be a bug, so I'd rather not risk introducing a bug here...

nilsvu replied:
why would prefixed vars be a bug? The system could choose to use prefixed vars for `variables_tag`. Everything works the same, it's just a label.

nilsdeppe replied:
Yes, on a quantity that's fundamental and shouldn't have one...

nilsvu commented on src/Evolution/Initialization/NonconservativeSystem.hpp:
```diff
@@ -0,0 +1,84 @@
+    static constexpr size_t dim = system::volume_dim;
+    using variables_tag = typename system::variables_tag;
+    using simple_tags = db::AddSimpleTags<variables_tag>;
+    using compute_tags = db::AddComputeTags<>;
```
remove line

nilsdeppe replied:
It's used below and makes the code easier to read. Leaving.

nilsdeppe committed:
Fix bug in building return types

Previously the return type building code would pass along all action lists
combined into one instead of only the action list of the current phase.

nilsdeppe commented:
@nilsleiffischer I've pushed new fixups and squashed the only ones. I had to add one new commit "Fix bug in building return types" to address one of your suggestions.

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -31,23 +31,24 @@ namespace detail {
 template <
     typename AddingAction, typename SimpleTag, MergePolicy Policy,
     typename DbTagsList,
     Requires<(Policy == MergePolicy::Error or Policy == MergePolicy::Ignore) and
```
What do you mean with "Ignore":
- Skip any SimpleTags that are already in the DataBox, not overwriting them.
- Skip only those SimpleTags that can't be compared for equivalence.

Currently you implement the second option, but I would expect the first since that's the opposite of what "Overwrite" does.

nilsdeppe replied:
Please see the documentation. This does exactly what it is documented to do.

nilsvu replied:
It does. What I meant is, I would expect that specifying "MergePolicy::Ignore" would actually ignore simple tags that have conflicts. Instead, it only ignores those that can't be compared. The current implementation doesn't allow for the case where you don't want the initialiser to overwrite the existing value.

nilsdeppe replied:
I don't think it's ever the correct answer to completely ignore the existing tags. That seems like it's asking for a bug so I'd rather not add an option that, at least to me currently, seems only useful for introducing bugs into the code. If you have a recommended alternative name I'm happy to rename the enum option.

nilsvu replied:
Perhaps make it `IgnoreIncomparable` or something (I know, it's about equivalence not comparability, but can't think of a better name). With the current implementation it's probably a pretty extreme edge-case, so the name won't be used a lot and can be long and descriptive. Most importantly, please fix the documentation for the enum case, which currently suggests this would actually ignore existing simple tags.

You could also just remove that case until we need it (if ever), and just error out if you can't compare the values.

nilsvu commented on src/Evolution/Initialization/Initialize.hpp:
```diff
@@ -54,9 +56,18 @@ struct RemoveOptionsAndTerminatePhase {
+    static_assert(std::is_same<tmpl::back<ActionList>,
+                               RemoveOptionsAndTerminatePhase>::value,
+                  "The last action in the initialization phase must be "
+                  "Initialization::Actions::RemoveOptionsAndTerminatePhase.");
```
It's the other way round: (If using this action,) `Initialization::Actions::RemoveOptionsAndTerminatePhase must be the last action in the initialization phase`. If the user doesn't use this action, e.g. removes option tags somewhere else, then some other action would end the initialization phase. To be clear, I'm only nitpicking on the grammar of this error message, not the functionality.

nilsdeppe replied:
I literally don't understand what your saying. I'm ignoring this.

nilsvu commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -101,6 +101,9 @@ struct ConservativeSystem {
+    ERROR(
+        "Could not find dependency 'Initialization::Tags::InitialTime' in "
+        "DataBox.");
     return {std::move(box)};
```
You can remove this line. I don't know what's the best-practice for this.

nilsdeppe replied:
It should be removed, I think :+1:

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,145 @@
+// Distributed under the MIT License.
```
Rename file to `MergeIntoDataBox` please

nilsvu commented on src/Evolution/Initialization/AddToDataBox.hpp:
```diff
@@ -0,0 +1,145 @@
+ * Compute tags that are not in the DataBox are added, ones that are in the
+ * DataBox already are ignored.
+ */
+template <typename AddingAction, typename ComputeTagsList,
```
What I meant in the other comment is to default `ComputeTagsList` to `tmp::list<>` here. Is that also what you meant, that you intentionally didn't do it? If so, why? Would be good to keep this as close to `db::create_from` as possible

nilsdeppe replied:
:man_shrugging: I'm leaving it.

nilsdeppe commented:
@nilsleiffischer rebased on develop and (hopefully) addressed all your comments.

nilsvu commented on docs/GroupDefs.hpp:
```diff
@@ -487,6 +487,12 @@
+/*!
+ * \defgroup InitializationGroup Initialization
+ * \brief Actions and metafunctions used for initialization of parallel
+ * components.
```
Just putting this here: Please consider what documentation people will need to get started with using the new initializers. Up to you to decide if more dox is needed now or defer that to later

nilsdeppe replied:
I think we should defer that to later since then there will be more tests that we can get snippets from. I think it should also be in `docs/DevGuide` so as to avoid making the `GroupDefs.hpp` too long (I know I've written docs in here in the past, and now regret it. Some have been moved).

nilsvu reviewed (commented):
Looks pretty much done to me. Please squash the fixups and get the commit history in the form you want it for merging (if it's not already). There's an unresolved thread here: https://github.com/sxs-collaboration/spectre/pull/1579#discussion_r297881777 If you end up changing code for that, please squash that directly.

Generally, please let the reviewers mark the threads as resolved if there's potential controversy. It's already a huge PR with >100 comments and hard to keep track of unresolved issues.

nilsdeppe commented:
@nilsleiffischer just did the rename.

nilsvu reviewed (dismissed):
For the record, this is missing tests for the initializers. I‘m adding tests for those initializers that generalize to elliptic systems in #1575. For the evolution-specific initializers I‘d be fine with adding tests now or later.

kidder commented on src/Evolution/Initialization/MergeIntoDataBox.hpp:
```diff
@@ -0,0 +1,148 @@
+ *
+ * How duplicate tags are handled depends on the `MergePolicy` passed. The
+ * default merge policy is to error if the simple tag being added does not have
+ * an equivalence operator and is being added again. `MergePolicy::Ignore` means
```
Ignore -> IgnoreIncomparable

kidder commented on tests/Unit/Evolution/Initialization/Test_MergeIntoDataBox.cpp:
```diff
@@ -0,0 +1,108 @@
+  CHECK(db::get<TagMultiplyByTwo<TagInt>>(box9) == 20);
+}
+}  // namespace TestAddToBox_detail
+
```
test the failure mode gives an ERROR

kidder commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -0,0 +1,48 @@
+  using type = std::vector<std::array<size_t, Dim>>;
+};
+
+template <size_t Dim>
```
why are there separate Initialization Tags? And in particular for Domain?

nilsvu replied:
FYI: I move them to `Domain/Tags.hpp` and `Evolution/Tags.hpp` in #1575

nilsdeppe replied:
I wanted to have it separate since this will need to be cleaned up once the Domain is in the global cache anyway.
